### PR TITLE
initialise github ci

### PR DESCRIPTION
**Pull request by chessai**
See the email I sent for context.

### CODE REVIEW

**Comment by Shimuuar** [comment]
I've updated actions:

    Fix gihub actions
    
    Now full range of supported GHC versions is tested on linux and latest 4 on
    windows/mac. Except 8.9.4 on windows which OOMs during building tests
    
     - Build is only triggered on PRs and push to master. Building vector ten times
       on every push seems excessive
    
     - Default shell is set to bash, otherwise windows build uses
       powershell. Hilarity ensues.
    
     - Build now first generates sdist tarball, unpacks it and only then builds
       package.
    
     - Doctests are fixed. Doctests rely on .ghc.environment files to load modules.
       --write-ghc-environment-files=always flag should be passed to build in order
       to generate them

Hopefully CI will work now

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+      matrix:
+        include:
+          # Linux
+          - { cabal: "3.2", os: ubuntu-latest,  ghc: "7.10.3" }
+          - { cabal: "3.2", os: ubuntu-latest,  ghc: "8.0.2"  }
+          - { cabal: "3.2", os: ubuntu-latest,  ghc: "8.2.2"  }
+          - { cabal: "3.2", os: ubuntu-latest,  ghc: "8.4.4"  }
+          - { cabal: "3.2", os: ubuntu-latest,  ghc: "8.6.5"  }
+          - { cabal: "3.2", os: ubuntu-latest,  ghc: "8.8.4"  }
+          - { cabal: "3.2", os: ubuntu-latest,  ghc: "8.10.2" }
```
Comment (Bodigrim) [inline .github/workflows/ci.yml:27]:
```suggestion
          - { cabal: "3.2", os: ubuntu-latest,  ghc: "8.10.3" }
```

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+          # Win
+          - { cabal: "3.2", os: windows-latest, ghc: "8.4.4"  }
+          - { cabal: "3.2", os: windows-latest, ghc: "8.6.5"  }
+          # OOM when building tests
+          # - { cabal: "3.2", os: windows-latest, ghc: "8.8.4"  }
+          - { cabal: "3.2", os: windows-latest, ghc: "8.10.1" }
```
Comment (Bodigrim) [inline .github/workflows/ci.yml:33]:
```suggestion
          - { cabal: "3.2", os: windows-latest, ghc: "8.10.3" }
```

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+          # MacOS
+          - { cabal: "3.2", os: macOS-latest,   ghc: "8.4.4"  }
+          - { cabal: "3.2", os: macOS-latest,   ghc: "8.6.5"  }
+          - { cabal: "3.2", os: macOS-latest,   ghc: "8.8.4"  }
+          - { cabal: "3.2", os: macOS-latest,   ghc: "8.10.1" }
```
Comment (Bodigrim) [inline .github/workflows/ci.yml:38]:
```suggestion
          - { cabal: "3.2", os: macOS-latest,   ghc: "8.10.3" }
```

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+      fail-fast: false
+
+    steps:
+    # ----------------
+    - uses: actions/checkout@v2
+      if: github.event.action == 'opened' || github.event.action == 'synchronize' || github.event.ref == 'refs/heads/master'
+    # ----------------
+    - uses: actions/setup-haskell@v1.1.4
```
Comment (Bodigrim) [inline .github/workflows/ci.yml:46]:
```suggestion
    - uses: haskell/actions/setup@v1
```

Comment (Shimuuar) [reply .github/workflows/ci.yml]: Thanks! I applied all suggestions

**Comment by Shimuuar** [comment]
At very least CI now works. I have no idea what to do with windows builds OOMing when building tests.

@Bodigrim @lehins should we merge this PR? At very least CI is not worse than it was

**Review by Bodigrim — APPROVED** [review]
Well done!

**Comment by lehins** [comment]
> I have no idea what to do with windows builds OOMing when building tests.

It is a known issue, I've been disabling ghc-8.10 on CI on windows since it was released

**Pull request merged**